### ISSUE TITLE
feature: add all/most known aggregators and attacks

### ISSUE THREAD

**Issue by srlt** [calicarpa/krum#14]
- including (where known) `upper_bound`, etc

**Comment by ArthurDanjou** [comment]
I've listed all the aggregators and attacks used in the list of the selected papers. We need to discuss to validate or not the implementation

## Attacks

| # | Name | Description | To Validate | Done |
|---|------|-------------|--------------|------|
| 1 | sign_flip / signflipping / SignFlipping | Invert gradient sign | | |
| 2 | label_flip / labelflipping / LabelFlipping | Flip labels on client side | | |
| 3 | mimic / Mimic | Mimic another worker's gradient | | |
| 4 | nan / NaN | Send NaN values | ☑ | |
| 5 | signflipping | Variant of sign flip | | |
| 6 | scaledSF / scaled_sign_flip | Scaled sign flipping | | |
| 7 | mimic_heuristic | Heuristic mimic | | |
| 8 | empire | Empire attack | ☑ | |
| 9 | anticge | Anti-CGE | | |
| 10 | identical_sparse | Identical sparse | | |
| 11 | identical | Identical gradients | | |
| 12 | dissension | Dissension | | |
| 13 | InnerProductManipulation / IPM | Scale mean by -tau | | |
| 14 | Optimal_InnerProductManipulation | Optimized IPM | | |
| 15 | ALittleIsEnough / ALIE / lie | Scale mean + tau * std | ☑ | |
| 16 | Optimal_ALittleIsEnough | Optimized ALIE | ☑ | |
| 17 | Inf | Send infinity | ☑ | | |
| 18 | Gaussian | Random gaussian noise | | |
| 19 | zero | Send zeros | | |
| 20 | noise | Add random noise | | |
| 21 | random | Random attack | | |
| 22 | byzMean | Byzantine mean | | |
| 23 | min_max / minmax | Min-max attack | | |
| 24 | min_sum | Min-sum attack | | |
| 25 | adaptive_std | Adaptive std attack | | |
| 26 | adaptive_sign | Adaptive sign attack | | |
| 27 | adaptive_uv | Adaptive UV attack | | |

## Aggregators

| # | Name | Description | To Validate | Done |
|---|------|-------------|--------------|------|
| 1 | average / Mean / Average | Simple arithmetic mean | ☑ | |
| 2 | median / Median | Coordinate-wise median | ☑ | |
| 3 | trmean / TrMean / trimmed_mean | Trimmed mean | ☑ | |
| 4 | krum / Krum | Krum aggregator | | |
| 5 | multikrum / MultiKrum / Multi-Krum | Multi-Krum | ☑ | |
| 6 | bulyan / Bulyan | Bulyan | ☑ | |
| 7 | geometric_median / GeoMed / GeometricMedian | Geometric median | | |
| 8 | dnc / DnC / divide_conquer | Divide and conquer | | |
| 9 | signguard / SignGuard | Sign guard | | |
| 10 | rfa / RFA | Robust federated aggregation | | |
| 11 | mom / MoM | Momentum-based | | |
| 12 | mva / MVA | Minimum variance averaging | | |
| 13 | mea / MEA | Mean estimation | | |
| 14 | cva / CVA | Coordinate-wise vector agreement | | |
| 15 | iter_cva / Iter_CVA | Iterative CVA | | |
| 16 | cenna / CENNA | Cross-entropy neural network aggregator | | |
| 17 | bucketing | Bucketing | | |
| 18 | bucketing_stress | Bucketing stress | | |
| 19 | brute / Brute | ~Brute force~ (= MDA) | | |
| 20 | aksel | Aksel | ☑ | |
| 21 | cge / CGE | CGE | | |
| 22 | centeredclip / CenteredClipping | Centered clipping | | |
| 23 | filterL2 | L2 filtering | | |
| 24 | krum_pseudo | Krum pseudo | | |
| 25 | multiKrum_pseudo | Multi-Krum pseudo | | |
| 26 | ios | IOS | | |
| 27 | robust_gossip | Robust gossip | | |
| 28 | MDA | Minimum Diameter Averaging | ☑ | |
| 29 | MoNNA | MoNNA | | |
| 30 | Meamed | Mean around median | ☑ | |
| 31 | CAF | Covariance-bound Agnostic Filter | | |
| 32 | SMEA | Smallest Maximum Eigenvalue Averaging | | |
| 33 | SignGuard-Sim | SignGuard simplified | | |
| 34 | SignGuard-Dist | SignGuard distant | | |

**Comment by moh** [comment]
Note: build base class first for attacks and GAR, think in a way to make it easy to use by researchers (adding new components ...)
